core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala

```diff
@@ -92,6 +92,9 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
   // The num of current max ExecutorId used to re-register appMaster
   @volatile protected var currentExecutorIdCounter = 0
 
+  // The set of executors we have on each host.
+  protected val hostToExecutors = new HashMap[String, HashSet[String]]
```

Comment:
What are you calling "a large cluster"?

Let's say you have 5000 executors. How expensive is this walk? Compared to sending messages over the network to kill the matching executors? Taking into account how often you actually have to call this method?